[ROM] Unify ROM solver

rubenzorrilla opened the pull request:
**📝 Description**
Similar to what we did in #9439 with the `RomAnalysisStage`, in this PR I'm unifying the ROM solvers in a new `RomSolver`. The dynamic inheritance mechanism is equivalent to that of #9439. In order to keep backwards compatibility I kept the old solvers (adding a deprecation warning) and created a `new_python_solvers_wrapper_rom.py` module. This module filters the base solver from which the ROM solver has to inherit.

In this regard I'd like to note that, if we had a unified name for all the Python solvers wrapper (e.g. `python_solvers_wrapper` instead of `python_solvers_wrapper_fluid`), the new ROM wrapper would be simplified and easier to maintain. In my opinion, it makes not so much sense to have different names for the Python solvers wrappers among applications if we follow the rule that one application should have only one "standard" wrapper (I said standard because of the adjoints). In any case this is to be discussed in an independent issue.   

This requires #9467 to be merged first.

rubenzorrilla committed:
Deprecate old solvers and wrapper

rubenzorrilla committed:
New generic ROM solver

rubenzorrilla committed:
Minor simplification

rubenzorrilla committed:
Minor simplification

rubenzorrilla committed:
Merge branch 'rom/unify-rom-solver' of https://github.com/KratosMultiphysics/Kratos into rom/unify-rom-solver

sunethwarna commented on applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_solver.py:
```diff
@@ -260,37 +260,39 @@ def GetMinimumBufferSize(self):
+        # Add the DOFs and reaction list to each node
+        KratosMultiphysics.VariableUtils.AddDofsList(dofs_with_reactions_list, self.main_model_part)
+
+        KratosMultiphysics.Logger.PrintInfo("::[ConvectionDiffusionSolver]:: ", "DOF's ADDED")
```
I would simplify this by using `GetDofsList` method, and adding reaction dofs afterwards.

sunethwarna commented on applications/FluidDynamicsApplication/python_scripts/adjoint_fluid_solver.py:
```diff
@@ -18,13 +18,25 @@ def __init__(self, model, settings):
+        dofs_to_add.append("ADJOINT_FLUID_VECTOR_1_Y")
+        dofs_to_add.append("ADJOINT_FLUID_VECTOR_1_Z")
+        dofs_to_add.append("ADJOINT_FLUID_SCALAR_1")
+        KratosMultiphysics.VariableUtils.AddDofsList(dofs_to_add, self.main_model_part)
```
Why can't we use the `GetDofsList` to add the dofs. If so, it will make this consistent. Do we need the `ADJOINT_FLUID_VECTOR_1_Z` always?. As far as I see, at least in FluidApp, `ADJOINT_FLUID_VECTOR_1_Z` dof is only checked and used in 3D.

rubenzorrilla replied:
Please see the discussion in #9467.

sunethwarna commented on applications/FluidDynamicsApplication/python_scripts/navier_stokes_compressible_solver.py:
```diff
@@ -115,11 +115,26 @@ def AddVariables(self):
+        dofs_and_reactions_to_add.append(["TOTAL_ENERGY", "REACTION_ENERGY"])
+        KratosMultiphysics.VariableUtils.AddDofsList(dofs_and_reactions_to_add, self.main_model_part)
+
+        KratosMultiphysics.Logger.PrintInfo(self.__class__.__name__, "Fluid solver DOFs added correctly.")
```
Again, could we use `GetDofsList`?

rubenzorrilla replied:
No. Please see the discussion in #9467.

sunethwarna commented on applications/FluidDynamicsApplication/python_scripts/fluid_solver.py:
```diff
@@ -66,13 +66,21 @@ def AddVariables(self):
+        """This function creates and returns a list with the DOFs defined in the conditions and elements specifications
+        Note that this requires the main_model_part to be already set, that is to say to have already performed the element substitution (see PrepareModelPart).
+        """
+        return KratosMultiphysics.SpecificationsUtilities.GetDofsListFromSpecifications(self.main_model_part)
```
If we are using this, then we don't need to define `GetDofsList` for each derrived class ryt? 

It is little bit confusing and inconsistent the current implementation. In the case if someone requests dofs list via `GetDofsList` before adding dofs, some solvers will give dofs correctly, some will give an empty list.

rubenzorrilla replied:
Please see the discussion in #9467.

sunethwarna commented on applications/RomApplication/python_scripts/new_python_solvers_wrapper_rom.py:
```diff
@@ -0,0 +1,61 @@
+        raise Exception("input is expected to be provided as a Kratos Model object")
+
+    if (type(solver_settings) != KratosMultiphysics.Parameters):
+        raise Exception("input is expected to be provided as a Kratos Parameters object")
```
Just suggesting, the more pythonic way is to use `isinstance` ryt?

rubenzorrilla replied:
Agree. I copied and pasted. Updating it. Thanks.

rubenzorrilla committed:
Using isinstance

philbucher commented on applications/RomApplication/python_scripts/new_python_solvers_wrapper_rom.py:
```diff
@@ -0,0 +1,61 @@
+import sys
```
I recommend coming up with a better name rather than prepending with `new_` 
remember how long we were stuck with the `new_linear_solver_factory` :sweat_smile:

rubenzorrilla replied:
:smile: Considering that this is a small research application I don't think we'll need to change keep backwards compatibility for a very long period of time. Indeed, the plan is to remove the old stuff ASAP. I'd keep the `new_` to constantly remind us to do it :smile:.

rubenzorrilla committed:
Fix syntax

rubenzorrilla committed:
Fix fix syntax

rubenzorrilla committed:
Merge branch 'master' into rom/unify-rom-solver

Rbravo555 commented on applications/RomApplication/python_scripts/rom_solver.py:
```diff
@@ -0,0 +1,58 @@
+            if len(nodal_unknowns) == 0:
+                solver_dofs_list = self.GetDofsList()
+                if not len(solver_dofs_list) == 0:
+                    self.settings["rom_settings"]["nodal_unknowns"].SetStringArray()
```
```suggestion
                    self.settings["rom_settings"]["nodal_unknowns"].SetStringArray(solver_dofs_list)
```

Rbravo555 replied:
I understand this line should tackle the rom_settings.nodal_unknowns being empty.

However, I have porpusely erased it `{
    "rom_settings": {
        "nodal_unknowns": [], ...`

But doing this change alone will not work if the user indeed enters an empty list. 

Line 56 in rom_analisys.py should also be changed

rubenzorrilla replied:
Nice catch! Indeed, while updating one of the examples in the Examples repo I realized that we need to slightly change the factory method in `rom_analysis.py`. I'd change in there.

Rbravo555 committed:
Update rom_analysis.py

making rom_analysis compatible with empty "nodal_unknowns" list in "rom_parameters"

Rbravo555 reviewed (commented):
LGTM now

Rbravo555 committed:
Update applications/RomApplication/python_scripts/rom_solver.py
